synedrion/src/cggmp21/entities.rs

```diff
@@ -92,40 +110,46 @@ pub struct KeyShareChange<P: SchemeParams, I: Ord> {
     pub(crate) phantom: PhantomData<P>,
 }
 
-/// The result of the Presigning protocol.
-#[derive(Debug, Clone)]
-pub(crate) struct PresigningData<P: SchemeParams, I> {
-    pub(crate) nonce: Scalar, // x-coordinate of $R$
-    /// An additive share of the ephemeral scalar.
-    pub(crate) ephemeral_scalar_share: Secret<Scalar>, // $k_i$
-    /// An additive share of `k * x` where `x` is the secret key.
-    pub(crate) product_share: Secret<Scalar>,
-
-    // Values generated during presigning,
-    // kept in case we need to generate a proof of correctness.
-    pub(crate) product_share_nonreduced: SecretSigned<<P::Paillier as PaillierParams>::Uint>,
-
-    // $K_i$.
-    pub(crate) cap_k: Ciphertext<P::Paillier>,
-
-    // The values for $j$, $j != i$.
-    pub(crate) values: BTreeMap<I, PresigningValues<P>>,
+impl<P: SchemeParams, I: Ord> PublicAuxInfos<P, I> {
+    pub(crate) fn as_map(&self) -> &BTreeMap<I, PublicAuxInfo<P>> {
+        &self.0
+    }
+
+    /// Returns a `PublicAuxInfos` object for the given subset of all parties.
+    pub fn subset(self, parties: &BTreeSet<I>) -> Result<Self, LocalError> {
+        if parties.iter().any(|party| !self.0.contains_key(party)) {
+            return Err(LocalError::new(
+                "`parties` must be a subset of all the available Aux Infos",
+            ));
+        }
+
+        let aux_infos = self.0.into_iter().filter(|(id, _aux)| parties.contains(id)).collect();
```

Comment:
Maybe we can skip the first iteration (lines 120-123) if we first construct the `aux_infos` and then compare the lengths: if `aux_infos` has fewer items than `parties` then return error? It would be optimizing for the optimistic case.

Reply:
Also: the case of nr of parties being wrong seems to not be covered by a test (not sure it's necessary to have one, just that I tried removing the check and it all ran green).

Reply:
Yeah, the test coverage is not full yet. We can add that in later PRs.